File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/pigeon_implement/BluetoothDelegateImpl.kt
package com.idosmart.pigeon_implement

import com.idosmart.pigeongen.api_bluetooth.*
import com.idosmart.protocol_sdk.*
import com.idosmart.tool.Help
import com.idosmart.model.*
import com.idosmart.enums.*
import com.idosmart.protocol_channel.plugin
import com.idosmart.protocol_channel.innerRunOnMainThread

internal class BluetoothDelegateImpl: BluetoothDelegate {

    companion object {
        @Volatile
        private var instance: BluetoothDelegateImpl? = null
        fun instance(): BluetoothDelegateImpl {
            return instance ?: synchronized(this) {
                instance ?: BluetoothDelegateImpl().also { instance = it }
            }
        }
    }

    private var delegate: IDOBleDelegate? = null

    private var dfuDelegate: IDODfuDelegate? = null

    private val mapOnlyOnceCallbackStateBt = mutableMapOf<String, Boolean>()

    var callbackBleDeviceModel: ((DeviceModel) -> Unit)? = null

    internal val bleDevice: IDOBleDeviceModel by lazy {
        IDOBleDeviceModel()
    }

    private val bluetoothState: IDOBluetoothStateModel by lazy {
        IDOBluetoothStateModel()
    }
    private val deviceStateModel: IDODeviceStateModel by lazy {
        IDODeviceStateModel()
    }

    /**
    添加代理对象
     */
    fun addDelegate(api:IDOBleDelegate?) {
        delegate = api
    }

    /**
     * 添加升级代理对象
     */
    fun addDfuDelegate(api: IDODfuDelegate?) {
        dfuDelegate = api
    }

    override fun writeState(state: WriteStateModel) {
//        val  bleWrite = IDOWriteStateModel()
//        bleWrite.state = state.state
//        bleWrite.uuid = state.uuid
//        bleWrite.macAddress = state.macAddress
//        bleWrite.type = IDOWriteType.values()[(state.type?.raw?:0)]
//        this.delegate?.writeState(bleWrite)
    }

    override fun receiveData(data: ReceiveData) {
        val receive = IDOReceiveData()
        receive.macAddress = data.macAddress ?: ""
        receive.spp = data.spp ?: false
        receive.data = data.data?: ByteArray(0)
        receive.uuid = data.uuid ?: ""
        receive.platform = data.platform ?: 0
        this.delegate?.receiveData(receive)
    }

    override fun scanResult(list: List<DeviceModel>?) {
        val deviceList = list?.let { Help.getDeviceList(it) }
        this.delegate?.scanResult(deviceList)
    }

    override fun bluetoothState(state: BluetoothStateModel) {
        bluetoothState.type = IDOBluetoothStateType.values()[state.type?.raw ?:0]
        bluetoothState.scanType = IDOBluetoothScanType.values()[state.scanType?.raw ?:0]
        this.delegate?.bluetoothState(bluetoothState)
    }

    override fun deviceState(state: DeviceStateModel) {
        deviceStateModel.uuid = state.uuid ?: ""
        deviceStateModel.macAddress = state.macAddress ?: ""
        deviceStateModel.state = IDODeviceStateType.values()[state.state?.raw ?:0]
        bleDevice.state = deviceStateModel.state
        if (bluetoothState.type == IDOBluetoothStateType.POWEREDOFF) {
            deviceStateModel.errorState = IDOConnectErrorType.BLUETOOTHOFF
        }else {
            deviceStateModel.errorState = IDOConnectErrorType.values()[state.errorState?.raw ?:0]
        }
        this.delegate?.deviceState(deviceStateModel)
    }

    override fun stateSPP(state: SPPStateModel) {
        val spp = IDOSppStateModel()
        spp.type = IDOSppStateType.values()[state.type?.raw?:0]
        this.delegate?.stateSPP(spp)
    }

    override fun writeSPPCompleteState(btMacAddress: String) {
        this.delegate?.writeSPPCompleteState(btMacAddress)
    }

    override fun dfuComplete() {
        this.dfuDelegate?.dfuComplete()
    }

    override fun dfuError(error: String) {
        this.dfuDelegate?.dfuError(error)
    }

    override fun dfuProgress(progress: Long) {
        this.dfuDelegate?.dfuProgress(progress.toInt())
    }

    override fun changeCurrentDevice(device: DeviceModel) {
        //println("device.btMacAddress = ${device.btMacAddress}")
        bleDevice.macAddress = device.macAddress
        bleDevice.updateValues(device)

        // 清理非当前设备的bt配对状态
        if (!device.macAddress.isNullOrEmpty()) {
            val curValue = mapOnlyOnceCallbackStateBt[device.macAddress]
            mapOnlyOnceCallbackStateBt.clear()
            if (curValue != null) {
                mapOnlyOnceCallbackStateBt[device.macAddress] = curValue
            }
        }else {
            mapOnlyOnceCallbackStateBt.clear()
        }
    }

    /** 监听当前设备属性变更（目前针对btMacAddress 内部使用） */
    override fun onCurrentDeviceAttrValChange(device: DeviceModel) {
        val state = IDODeviceStateType.values()[device.state?.raw ?:0]
        //println("device.btMacAddress2 = ${device.btMacAddress}")
        bleDevice.macAddress = device.macAddress
        bleDevice.updateValues(device)
        callbackBleDeviceModel?.invoke(device)
    }

    override fun stateBt(isPair: Boolean) {
        printLog("stateBt - isPair: $isPair")
        if (bleDevice.macAddress.isNullOrEmpty()) {
            printLog("stateBt - macAddress is empty, return")
            return
        }

        // 过滤重复状态回调（原生android执行bt配对，当配对失败时，系统存在多次回调）
        val macAddress = bleDevice.macAddress!!
        if (!mapOnlyOnceCallbackStateBt.contains(macAddress) ||
            mapOnlyOnceCallbackStateBt[macAddress] != isPair) {
            this.delegate?.stateBT(isPair)
            mapOnlyOnceCallbackStateBt[bleDevice.macAddress!!] = isPair
            printLog("stateBt - call stateBT delegate")
        }else {
            printLog("stateBt - call stateBT delegate, ignore")
        }
    }

    private fun printLog(msg: String) {
        innerRunOnMainThread {
            plugin.tool()?.logNative(msg) { }
        }
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/pigeon_implement/BluetoothDelegateImpl.swift
class BluetoothDelegateImpl: BluetoothDelegate {
    
    var callbackDeviceBleModel: ((DeviceModel) -> Void)?
    private var _deviceBleModel: DeviceModel?
    private var _lastDeviceStateModel: IDODeviceStateModel?
    private var _lastBluetoothStateModel: IDOBluetoothStateModel?
    static let shared = BluetoothDelegateImpl()
    
    weak var delegate: IDOBleDelegate?
    weak var delegateDfu: IDODfuDelegate?
    
    private init() {}
    func onCurrentDeviceAttrValChange(device: DeviceModel) throws {
        _deviceBleModel = device
        callbackDeviceBleModel?(_deviceBleModel!)
    }
    
    func changeCurrentDevice(device: DeviceModel) throws {
        _deviceBleModel = device
        callbackDeviceBleModel?(_deviceBleModel!)
    }
    
    func writeState(state: WriteStateModel) throws {
        //delegate?.writeState(state: state.toWriteStateModel())
    }
    
    func receiveData(data: ReceiveData) throws {
        delegate?.receiveData(data: data.toReceiveData())
    }
    
    func scanResult(list: [DeviceModel]?) throws {
        guard let items = list else {
            delegate?.scanResult(list: nil)
            return
        }
        let rsList = items.map { $0.toIDODeviceModel() }
        delegate?.scanResult(list: rsList)
    }
    
    func bluetoothState(state: BluetoothStateModel) throws {
        _lastBluetoothStateModel = state.toIDOBluetoothStateModel()
        delegate?.bluetoothState(state: _lastBluetoothStateModel!)
        if (_lastDeviceStateModel != nil && state.type == .poweredOff) {
            _lastDeviceStateModel?.state = .disconnected
            _lastDeviceStateModel?.errorState = .bluetoothOff
            delegate?.deviceState(state: _lastDeviceStateModel!)
        }
    }
    
    func deviceState(state: DeviceStateModel) throws {
        //_logNative("调用到 swift deviceState: \(state.state) delegate:\(delegate)")
        if (_lastDeviceStateModel == nil) {
            _lastDeviceStateModel = state.toDeviceStateModel()
        }
        _lastDeviceStateModel?.uuid = state.uuid
        _lastDeviceStateModel?.macAddress = state.macAddress
        _lastDeviceStateModel?.state = IDODeviceStateType(rawValue: state.state!.rawValue)!
        if (_lastBluetoothStateModel?.type != .poweredOff) {
            _lastDeviceStateModel?.errorState = IDOConnectErrorType(rawValue: state.errorState!.rawValue)!
        }
        delegate?.deviceState(state: _lastDeviceStateModel!)
    }
    
    func stateSPP(state: SPPStateModel) throws { }
    
    func writeSPPCompleteState(btMacAddress: String) throws { }
    
    func dfuComplete() throws {
        delegateDfu?.dfuComplete()
    }
    
    func dfuError(error: String) throws {
        delegateDfu?.dfuError(error: error)
    }
    
    func dfuProgress(progress: Int64) throws {
        delegateDfu?.dfuProgress(progress: Int(progress))
    }

    // ios使用不到
    func stateBt(isPair: Bool) throws { }
}


// MARK: -

extension IDODfuConfig {
    func toDfuConfig() -> DfuConfig {
        return DfuConfig(
            filePath: filePath,
            uuid: uuid,
            macAddress: macAddress,
            deviceId: deviceId,
            platform: platform.int64,
            isDeviceSupportPairedWithPhoneSystem: isDeviceSupportPairedWithPhoneSystem,
            prn: prn.int64,
            isNeedReOpenBluetoothSwitchIfFailed: isNeedReOpenBluetoothSwitchIfFailed,
            maxRetryTime: maxRetryTime.int64,
            isNeedAuth: isNeedAuth,
            otaWorkMode: otaWorkMode.int64
        )
    }
}

extension SPPStateModel {
    func toSPPStateModel() -> IDOSppStateModel {
        let _type = IDOSppStateType(rawValue: type!.rawValue)!
        return IDOSppStateModel(type: _type)
    }
}

extension DeviceStateModel {
    func toDeviceStateModel() -> IDODeviceStateModel {
        let errorState = IDOConnectErrorType(rawValue: errorState!.rawValue)!
        let _state = IDODeviceStateType(rawValue: state!.rawValue)!
        return IDODeviceStateModel(uuid: uuid, macAddress: macAddress, state: _state, errorState: errorState)
    }
}

extension BluetoothStateModel {
    func toIDOBluetoothStateModel() -> IDOBluetoothStateModel {
        var _type: IDOBluetoothStateType = .unknown
        if (type != nil) {
            _type = IDOBluetoothStateType(rawValue: type!.rawValue)!
        }
        var _scanType: IDOBluetoothScanType = .stop
        if (scanType != nil) {
            _scanType = IDOBluetoothScanType(rawValue: scanType!.rawValue)!
        }
        return IDOBluetoothStateModel(type: _type, scanType: _scanType)
    }
}

// extension IDOBluetoothStateModel {
//    func toBluetoothStateModel() -> BluetoothStateModel {
//        let type = BluetoothStateType(rawValue: type!.rawValue)
//        return BluetoothStateModel(type: type)
//    }
// }

extension ReceiveData {
    func toReceiveData() -> IDOReceiveData {
        return IDOReceiveData(data: data?.data, uuid: uuid, macAddress: macAddress, spp: spp ?? false, platform: platform?.int ?? 0)
    }
}

extension WriteStateModel {
    func toWriteStateModel() -> IDOWriteStateModel {
        let _type = IDOWriteType(rawValue: type!.rawValue)!
        return IDOWriteStateModel(state: state ?? false, uuid: uuid, macAddress: macAddress, type: _type)
    }
}

extension DeviceModel {
    func toIDODeviceModel() -> IDODeviceModel {
        let _state = IDODeviceStateType(rawValue: state!.rawValue)!
        return IDODeviceModel(rssi: rssi?.int ?? 0,
                              name: name,
                              state: _state,
                              uuid: uuid,
                              macAddress: macAddress,
                              otaMacAddress: otaMacAddress,
                              btMacAddress: btMacAddress,
                              deviceId: deviceId?.int ?? 0,
                              deviceType: deviceType?.int ?? 0,
                              isOta: isOta ?? false,
                              isTlwOta: isTlwOta ?? false,
                              bltVersion: bltVersion?.int ?? 0,
                              isPair: isPair ?? false,
                              platform: platform?.int ?? 0)
    }
}

extension IDODeviceModel {
    func toDeviceModel() -> DeviceModel {
        let _state = DeviceStateType(rawValue: state.rawValue)
        return DeviceModel(rssi: rssi.int64,
                           name: name,
                           state: _state,
                           uuid: uuid,
                           macAddress: macAddress,
                           otaMacAddress: otaMacAddress,
                           btMacAddress: btMacAddress,
                           deviceId: deviceId.int64,
                           deviceType: deviceType.int64,
                           isOta: isOta,
                           isTlwOta: isTlwOta,
                           bltVersion: bltVersion.int64,
                           isPair: isPair,
                           platform: platform.int64)
    }
}
